File: MY_REPOS/Lambda-Resource-Static-Assets/2-resources/_External-learning-resources/00-Javascript/Object-oriented-programming-for-JavaScript/Module 2/B04710_CodeBundle/Chapter 3/B04170_03_Python_Draft_01.cs
class SmoothFoxTerrier
{
    public static string Family = "Terrier";
    public static string AreaOfOrigin = "England";
    public static int Energy = 10;
    public static int ColdTolerance = 8;
    public static int HeatTolerance = 8;

    public string Name;
    public int WatchdogAbility;
    public string FavoriteToy;

    public SmoothFoxTerrier(string name, int watchdogAbility, string favoriteToy)
    {
        this.Name = name;
        this.WatchdogAbility = watchdogAbility;
        this.FavoriteToy = favoriteToy;
    }
}



Debug.WriteLine(SmoothFoxTerrier.Family);



SmoothFoxTerrier.Energy = 8;



class SmoothFoxTerrier
{
    public static string Family = "Terrier";
    public static string AreaOfOrigin = "England";
    public const int Energy = 10;
    public static int ColdTolerance = 8;
    public static int HeatTolerance = 8;

    public string Name;
    public int WatchdogAbility;
    public string FavoriteToy;

    public SmoothFoxTerrier(string name, int watchdogAbility, string favoriteToy)
    {
        this.Name = name;
        this.WatchdogAbility = watchdogAbility;
        this.FavoriteToy = favoriteToy;
    }
}



class SmoothFoxTerrier
{
    public static string Family = "Terrier";
    public static string AreaOfOrigin = "England";
    public static int Energy = 10;
    public static int ColdTolerance = 8;
    public static int HeatTolerance = 8;

    protected string _name;
    protected int _watchdogAbility;
    protected string _favoriteToy;

    public SmoothFoxTerrier(string name, int watchdogAbility, string favoriteToy)
    {
        this._name = name;
        this._watchdogAbility = watchdogAbility;
        this._favoriteToy = favoriteToy;
    }
}



class SmoothFoxTerrier
{
    public static string Family = "Terrier";
    public static string AreaOfOrigin = "England";
    public static int Energy = 10;
    public static int ColdTolerance = 8;
    public static int HeatTolerance = 8;

    private string _name;
    private int _watchdogAbility;
    private string _favoriteToy;

    public SmoothFoxTerrier(string name, int watchdogAbility, string favoriteToy)
    {
        this._name = name;
        this._watchdogAbility = watchdogAbility;
        this._favoriteToy = favoriteToy;
    }
}



class SmoothFoxTerrier
{
    public static string Family = "Terrier";
    public static string AreaOfOrigin = "England";
    public static int Energy = 10;
    public static int ColdTolerance = 8;
    public static int HeatTolerance = 8;

    private string _name;
    private int _watchdogAbility;
    private string _favoriteToy;

    public string Name 
    {
        get
        {
            return _name;
        }
    }

    public SmoothFoxTerrier(string name, int watchdogAbility, string favoriteToy)
    {
        this._name = name;
        this._watchdogAbility = watchdogAbility;
        this._favoriteToy = favoriteToy;
    }
}



var jerry = new SmoothFoxTerrier("Jerry", 7, "Boomerang");
jerry.Name = "Tom";



class SmoothFoxTerrier
{
    public static string Family = "Terrier";
    public static string AreaOfOrigin = "England";
    public static int Energy = 10;
    public static int ColdTolerance = 8;
    public static int HeatTolerance = 8;

    private string _name;
    private int _watchdogAbility;
    private string _favoriteToy;

    public string Name 
    {
        get
        {
            return _name;
        }
    }

    public string FavoriteToy
    {
        get 
        {
            return this._favoriteToy;
        }

        set 
        {
            this._favoriteToy = value;
        }
    }

    public SmoothFoxTerrier(string name, int watchdogAbility, string favoriteToy)
    {
        this._name = name;
        this._watchdogAbility = watchdogAbility;
        this._favoriteToy = favoriteToy;
    }
}



class SmoothFoxTerrier
{
    public static string Family = "Terrier";
    public static string AreaOfOrigin = "England";
    public static int Energy = 10;
    public static int ColdTolerance = 8;
    public static int HeatTolerance = 8;

    private string _name;
    private int _watchdogAbility;
    private string _favoriteToy;

    public string Name 
    {
        get
        {
            return _name;
        }
    }

    public string FavoriteToy
    {
        get 
        {
            return this._favoriteToy;
        }

        set 
        {
            this._favoriteToy = value;
        }
    }

    public int WatchdogAbility
    {
        get
        {
            return this._watchdogAbility;
        }
        set 
        {
            if (value < 0) {
                this._watchdogAbility = 0;
            } else if (value > 10) {
                this._watchdogAbility = 10;
            } else {
                this._watchdogAbility = value;
            }
        }
    }

    public SmoothFoxTerrier(string name, int watchdogAbility, string favoriteToy)
    {
        this._name = name;
        this._watchdogAbility = watchdogAbility;
        this._favoriteToy = favoriteToy;
    }
}



var tom = new SmoothFoxTerrier("Tom", 8, "Boomerang");
tom.WatchdogAbility = -9;
Console.WriteLine(tom.WatchdogAbility);
tom.WatchdogAbility = 52;
Console.WriteLine(tom.WatchdogAbility);
tom.WatchdogAbility = 9;
Console.WriteLine(tom.WatchdogAbility);



class SmoothFoxTerrier
{
    public static string Family = "Terrier";
    public static string AreaOfOrigin = "England";
    public static int Energy = 10;
    public static int ColdTolerance = 8;
    public static int HeatTolerance = 8;

    private string _name;
    private int _watchdogAbility;
    private string _favoriteToy;

    public string Name 
    {
        get
        {
            return _name;
        }
    }

    public string FavoriteToy
    {
        get 
        {
            return this._favoriteToy;
        }

        set 
        {
            this._favoriteToy = value;
        }
    }

    public int WatchdogAbility
    {
        get
        {
            return this._watchdogAbility;
        }
        set 
        {
            if (value < 0) {
                this._watchdogAbility = 0;
            } else if (value > 10) {
                this._watchdogAbility = 10;
            } else {
                this._watchdogAbility = value;
            }
        }
    }

    public int ProtectionScore
    {
        get 
        {
            return (int)Math.Floor ((this._watchdogAbility + SmoothFoxTerrier.ColdTolerance + SmoothFoxTerrier.HeatTolerance) / 3d);
        }
    }

    public SmoothFoxTerrier(string name, int watchdogAbility, string favoriteToy)
    {
        this._name = name;
        this._watchdogAbility = watchdogAbility;
        this._favoriteToy = favoriteToy;
    }
}



var laura = new SmoothFoxTerrier("Laura", "Old sneakers", 9);
console.WriteLine(laura.ProtectionScore);



class SmoothFoxTerrier
{
    public static string Family = "Terrier";
    public static string AreaOfOrigin = "England";
    public const int Energy = 10;
    public static int ColdTolerance = 8;
    public static int HeatTolerance = 8;

    private string _name;
    private int _watchdogAbility;
    private string _favoriteToy;

    public string Name 
    {
        get
        {
            return this._name;
        }
    }

    public string FavoriteToy
    {
        get 
        {
            return this._favoriteToy;
        }

        set 
        {
            this._favoriteToy = value;
        }
    }

    public int WatchdogAbility
    {
        get
        {
            return this._watchdogAbility;
        }
        set 
        {
            if (value < 0) {
                this._watchdogAbility = 0;
            } else if (value > 10) {
                this._watchdogAbility = 10;
            } else {
                this._watchdogAbility = value;
            }
        }
    }

    public int ProtectionScore
    {
        get 
        {
            return Math.Floor ((this._watchdogAbility + SmoothFoxTerrier.ColdTolerance + SmoothFoxTerrier.HeatTolerance) / 3);
        }
    }

    public SmoothFoxTerrier(string name, int watchdogAbility, string favoriteToy)
    {
        this._name = name;
        this._watchdogAbility = watchdogAbility;
        this._favoriteToy = favoriteToy;
    }
}



class SmoothFoxTerrier
{
    public static string Family = "Terrier";
    public static string AreaOfOrigin = "England";
    public const int Energy = 10;
    public static int ColdTolerance = 8;
    public static int HeatTolerance = 8;

    private string _name;
    private int _watchdogAbility;

    public string Name 
    {
        get
        {
            return this._name;
        }
    }

    public string FavoriteToy { get; set; } 

    public int WatchdogAbility
    {
        get
        {
            return this._watchdogAbility;
        }
        set 
        {
            if (value < 0) {
                this._watchdogAbility = 0;
            } else if (value > 10) {
                this._watchdogAbility = 10;
            } else {
                this._watchdogAbility = value;
            }
        }
    }

    public int ProtectionScore
    {
        get 
        {
            return Math.Floor ((this._watchdogAbility + SmoothFoxTerrier.ColdTolerance + SmoothFoxTerrier.HeatTolerance) / 3);
        }
    }

    public SmoothFoxTerrier(string name, int watchdogAbility, string favoriteToy)
    {
        this._name = name;
        this._watchdogAbility = watchdogAbility;
        this.FavoriteToy = favoriteToy;
    }
}



class SmoothFoxTerrier
{
    public static string Family = "Terrier";
    public static string AreaOfOrigin = "England";
    public const int Energy = 10;
    public static int ColdTolerance = 8;
    public static int HeatTolerance = 8;

    private int _watchdogAbility;

    public string Name { get; private set; }

    public string FavoriteToy { get; set; } 

    public int WatchdogAbility
    {
        get
        {
            return this._watchdogAbility;
        }
        set 
        {
            if (value < 0) {
                this._watchdogAbility = 0;
            } else if (value > 10) {
                this._watchdogAbility = 10;
            } else {
                this._watchdogAbility = value;
            }
        }
    }

    public int ProtectionScore
    {
        get 
        {
            return Math.Floor ((this._watchdogAbility + SmoothFoxTerrier.ColdTolerance + SmoothFoxTerrier.HeatTolerance) / 3);
        }
    }

    public SmoothFoxTerrier(string name, int watchdogAbility, string favoriteToy)
    {
        this.Name = name;
        this._watchdogAbility = watchdogAbility;
        this.FavoriteToy = favoriteToy;
    }
}



class MutableVector3D
{
    public double X { get; set; }
    public double Y { get; set; }
    public double Z { get; set; }

    public void Sum(double deltaX, double deltaY, double deltaZ)
    {
        this.X += deltaX;
        this.Y += deltaY;
        this.Z += deltaZ;
    }

    public MutableVector3D(double x, double y, double z)
    {
        this.X = x;
        this.Y = y;
        this.Z = z;
    }
}



class MutableVector3D
{
    public double X { get; set; }
    public double Y { get; set; }
    public double Z { get; set; }

    public void Sum(double deltaX, double deltaY, double deltaZ)
    {
        this.X += deltaX;
        this.Y += deltaY;
        this.Z += deltaZ;
    }

    public MutableVector3D(double x, double y, double z)
    {
        this.X = x;
        this.Y = y;
        this.Z = z;
    }

    public static MutableVector3D OriginVector()
    {
        return new MutableVector3D(0, 0, 0);
    }
}



var mutableVector3D = MutableVector3D.OriginVector();
mutableVector3D.Sum(5, 10, 15);
Console.WriteLine(mutableVector3D.X, mutableVector3D.Y, mutableVector3D.Z)



class ImmutableVector3D
{
    public double X { get; private set; }
    public double Y { get; private set; }
    public double Z { get; private set; }

    public ImmutableVector3D Sum(double deltaX, double deltaY, double deltaZ)
    {
        return new ImmutableVector3D (
            this.X + deltaX,
            this.Y + deltaY,
            this.Z + deltaZ);
    }

    public ImmutableVector3D(double x, double y, double z)
    {
        this.X = x;
        this.Y = y;
        this.Z = z;
    }

    public static ImmutableVector3D EqualElementsVector(double initialValue)
    {
        return new ImmutableVector3D(initialValue, initialValue, initialValue);
    }

    public static ImmutableVector3D OriginVector()
    {
        return ImmutableVector3D.EqualElementsVector(0);
    }
}



var vector0 = ImmutableVector3D.OriginVector();
var vector1 = vector0.Sum(5, 10, 15);
Console.WriteLine(vector1.X, vector1.Y, vector1.Z);

File: MY_REPOS/Lambda-Resource-Static-Assets/2-resources/_External-learning-resources/00-Javascript/Object-oriented-programming-for-JavaScript/Module 2/B04710_CodeBundle/Chapter 3/B04170_03_Python_Draft_01.py
class TibetanSpaniel:
    family = "Companion, herding"
    area_of_origin = "Tibet"
    learning_rate = 9
    obedience = 3
    problem_solving = 8

    def __init__(self, name, favorite_toy, watchdog_ability):
        self.name = name
        self.watchdog_ability = watchdog_ability
        self.favorite_toy = favorite_toy



print(TibetanSpaniel.family)



brian = TibetanSpaniel("Brian", "Talking Minion", 4)
print(brian.family)



TibetanSpaniel.obedience = 4



brian.obedience = 8
print(type(brian).obedience)
print(TibetanSpaniel.obedience)
print(brian.obedience)


class TibetanSpaniel:
    family = "Companion, herding"
    area_of_origin = "Tibet"
    learning_rate = 9
    obedience = 3
    problem_solving = 8

    def __init__(self, name, favorite_toy, watchdog_ability):
        self._name = name
        self._watchdog_ability = watchdog_ability
        self._favorite_toy = favorite_toy



class TibetanSpaniel:
    family = "Companion, herding"
    area_of_origin = "Tibet"
    learning_rate = 9
    obedience = 3
    problem_solving = 8

def __init__(self, name, favorite_toy, watchdog_ability):
    self.__name = name
    self.__watchdog_ability = watchdog_ability
    self.__favorite_toy = favorite_toy



class TibetanSpaniel:
    family = "Companion, herding"
    area_of_origin = "Tibet"
    learning_rate = 9
    obedience = 3
    problem_solving = 8

    def __init__(self, name, favorite_toy, watchdog_ability):
        self.__name = name
        self.__watchdog_ability = watchdog_ability
        self.__favorite_toy = favorite_toy

    @property
    def name(self):
        return self.__name



merlin = TibetanSpaniel("Merlin", "Talking Smurf", 6)
merlin.name = "brian"



class TibetanSpaniel:
    family = "Companion, herding"
    area_of_origin = "Tibet"
    learning_rate = 9
    obedience = 3
    problem_solving = 8

    def __init__(self, name, favorite_toy, watchdog_ability):
        self.__name = name
        self.__watchdog_ability = watchdog_ability
        self.__favorite_toy = favorite_toy

    @property
    def name(self):
        return self.__name

    @property
    def favorite_toy(self):
        return self.__favorite_toy

    @favorite_toy.setter
    def favorite_toy(self, favorite_toy):
        self.__favorite_toy = favorite_toy



class TibetanSpaniel:
    family = "Companion, herding"
    area_of_origin = "Tibet"
    learning_rate = 9
    obedience = 3
    problem_solving = 8

    def __init__(self, name, favorite_toy, watchdog_ability):
        self.__name = name
        self.__watchdog_ability = watchdog_ability
        self.__favorite_toy = favorite_toy

    @property
    def name(self):
        return self.__name

    @property
    def favorite_toy(self):
        return self.__favorite_toy

    @favorite_toy.setter
    def favorite_toy(self, favorite_toy):
        self.__favorite_toy = favorite_toy

    @property
    def watchdog_ability(self):
        return self.__watchdog_ability

    @watchdog_ability.setter
    def watchdog_ability(self, watchdog_ability):
        if watchdog_ability < 0:
            self.__watchdog_ability = 0
        elif watchdog_ability > 10:
            self.__watchdog_ability = 10
        else:
            self.__watchdog_ability = watchdog_ability



hugo = TibetanSpaniel("Hugo", "Tennis ball", 7)
hugo.watchdog_ability = -3
print(hugo.watchdog_ability)
hugo.watchdog_ability = 30
print(hugo.watchdog_ability)
hugo.watchdog_ability = 8
print(hugo.watchdog_ability)



class TibetanSpaniel:
    family = "Companion, herding"
    area_of_origin = "Tibet"
    learning_rate = 9
    obedience = 3
    problem_solving = 8

    def __init__(self, name, favorite_toy, watchdog_ability):
        self.__name = name
        self.__watchdog_ability = watchdog_ability
        self.__favorite_toy = favorite_toy

    @property
    def name(self):
        return self.__name

    @property
    def favorite_toy(self):
        return self.__favorite_toy

    @favorite_toy.setter
    def favorite_toy(self, favorite_toy):
        self.__favorite_toy = favorite_toy

    @property
    def watchdog_ability(self):
        return self.__watchdog_ability

    @watchdog_ability.setter
    def watchdog_ability(self, watchdog_ability):
        if watchdog_ability < 0:
            self.__watchdog_ability = 0
        elif watchdog_ability > 10:
            self.__watchdog_ability = 10
        else:
            self.__watchdog_ability = watchdog_ability
    
    @property
    def protection_score(self):
        return math.floor((self.__watchdog_ability + type(self).learning_rate + type(self).problem_solving) / 3)



cole = TibetanSpaniel("Cole", "Soccer ball", 4)
print(cole.protection_score)



class TibetanSpaniel:
    family = "Companion, herding"
    area_of_origin = "Tibet"
    learning_rate = 9
    obedience = 3
    problem_solving = 8

    def __init__(self, name, favorite_toy, watchdog_ability):
        self.__name = name
        self.__watchdog_ability = watchdog_ability
        self.__favorite_toy = favorite_toy

    @property
    def name(self):
        return self.__name

    @property
    def favorite_toy(self):
        return self.__favorite_toy

    @favorite_toy.setter
    def favorite_toy(self, favorite_toy):
        self.__favorite_toy = favorite_toy

    @property
    def watchdog_ability(self):
        return self.__watchdog_ability

    @watchdog_ability.setter
    def watchdog_ability(self, watchdog_ability):
        if watchdog_ability < 0:
            self.__watchdog_ability = 0
        elif watchdog_ability > 10:
            self.__watchdog_ability = 10
        else:
            self.__watchdog_ability = watchdog_ability

    @property
    def protection_score(self):
        return math.floor((self.__watchdog_ability + type(self).learning_rate + type(self).problem_solving) / 3)



class MutableVector3D:
    def __init__(self, x, y, z):
        self.__x = x
        self.__y = y
        self.__z = z

    def sum(self, delta_x, delta_y, delta_z):
        self.__x += delta_x
        self.__y += delta_y
        self.__z += delta_z

    @property
    def x(self):
        return self.__x

    @x.setter
    def x(self, x):
        self.__x = x

    @property
    def y(self):
        return self.__y

    @y.setter
    def y(self, y):
        self.__y = y

    @property
    def z(self):
        return self.__z

    @z.setter
    def z(self, z):
        self.__z = z



class MutableVector3D:
    def __init__(self, x, y, z):
        self.__x = x
        self.__y = y
        self.__z = z

    def sum(self, delta_x, delta_y, delta_z):
        self.__x += delta_x
        self.__y += delta_y
        self.__z += delta_z

    @property
    def x(self):
        return self.__x

    @x.setter
    def x(self, x):
        self.__x = x

    @property
    def y(self):
        return self.__y

    @y.setter
    def y(self, y):
        self.__y = y

    @property
    def z(self):
        return self.__z

    @z.setter
    def z(self, z):
        self.__z = z

    @classmethod
    def origin_vector(cls):
        return cls(0, 0, 0)



mutableVector3D = MutableVector3D.origin_vector()
mutableVector3D.sum(5, 10, 15)
print(mutableVector3D.x, mutableVector3D.y, mutableVector3D.z)



class ImmutableVector3D:
    def __init__(self, x, y, z):
        self.__x = x
        self.__y = y
        self.__z = z

    def sum(self, delta_x, delta_y, delta_z):
        return type(self)(self.__x + delta_x, self.__y + delta_y, self.__z + delta_z)

    @property
    def x(self):
        return self.__x

    @property
    def y(self):
        return self.__y

    @property
    def z(self):
        return self.__z

    @classmethod
    def equal_elements_vector(cls, initial_value):
        return cls(initial_value, initial_value, initial_value)

    @classmethod
    def origin_vector(cls):
        return cls.equal_elements_vector(0)



vector0 = ImmutableVector3D.origin_vector()
vector1 = vector0.sum(5, 10, 15)
print(vector1.x, vector1.y, vector1.z)
